tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/caches/Cache.scala

```diff
@@ -0,0 +1,177 @@
+
+  /**
+   * Compute a target value and context update, based on the source value, but do not save the result.
+   *
+   * @param context
+   *   the context before creating a new value
+   * @param srcValue
+   *   a source value
+   * @return
+   *   a target value that can be cached, and the new context
+   */
+  protected def create(context: ContextT, srcValue: SourceT): (ContextT, TargetT)
+
+  /**
+   * Get a previously cached value for a given source value, or return the previously cached one. Whenever a new value
+   * is created, it is cached. The cached value can be later removed by pop.
+   *
+   * @param srcValue
+   *   a source value
+   * @return
+   *   a target value
+   */
+  def getOrCreate(context: ContextT, srcValue: SourceT): (ContextT, TargetT) =
+    if (_cache.contains(srcValue)) {
+      (context, _cache(srcValue)._1)
+    } else {
+      // introduce a new constant
+      val (newContext, targetValue) = create(context, srcValue)
+      addToCache(srcValue, targetValue)
+      (newContext, targetValue)
+    }
+
+  /**
+   * Add a value to the cache.
+   *
+   * @param source
+   *   the source value
+   * @param target
+   *   the cached target value
+   */
+  protected def addToCache(source: SourceT, target: TargetT): Unit = {
+    _cache += source -> (target, level)
+    _reverseCache += target -> (source, level)
+  }
+
+  /**
+   * Get a previously cached value for a given source value, if there is one. Otherwise, return none.
+   *
+   * @param srcValue
+   *   a source value
+   * @return
+   *   Some(result), or None
+   */
+  def get(srcValue: SourceT): Option[TargetT] = cache.get(srcValue).map(_._1)
+
+  /**
+   * Find the key that was used to create a given value.
+   * @param value
+   *   a value, for which the key should be found
+   * @return
+   *   the key, if exists
+   */
+  def findKey(value: TargetT): Option[SourceT] = reverseCache.get(value).map(_._1)
+
+  /**
+   * Get the current level, that is the difference between the number of pushes and pops made so far.
+   *
+   * @return
+   *   the current level, always non-negative.
+   */
+  override def level: Int = _level
+
+  /**
+   * Save the current state and push it on the stack for a later recovery with pop. Increases level by 1.
+   */
+  override def snapshot(): Unit = _level += 1
+
+  /**
+   * Discard all entries added at the current level. Decreases level by 1.
+   *
+   * Importantly, pop may be called multiple times and thus it is not sufficient to save only the latest state.
+   */
+  override def revert(): Unit = revert(1)
+
+  /**
+   * Pop n times.
+   * @param n
+   *   pop n times, if n > 0, otherwise, do nothing
+   */
+  override def revert(n: Int): Unit = {
+    require(level >= n, s"Can't pop $n levels from a cache of level $level.")
+    _level -= n
+
+    def isOld[A, B](mapEntry: (A, (B, Int))): Boolean =
+      mapEntry._2._2 <= level
+
+    _cache = cache.filter(isOld)
+    _reverseCache = reverseCache.filter(isOld)
```

Comment:
`isOld` is a confusing name here. how about:

```suggestion
    def isRetained[A, B](mapEntry: (A, (B, Int))): Boolean =
      mapEntry._2._2 <= level

    _cache = cache.filter(isRetained)
    _reverseCache = reverseCache.filter(isRetained)
```